lib/create-plugin.js

```diff
@@ -3,11 +3,126 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
- 'use strict'
+'use strict'
 
-function createPlugin() {
+function createPlugin(instrumentationApi) {
   return {
+    requestDidStart() {
+      const longestResolvedPath = {
+        depth: 0,
+        formatted: null
+      }
 
+      return {
+        executionDidStart: () => ({
+          willResolveField({info}) {
+            const pathArray = flattenToArray(info.path)
+            const formattedPath = pathArray.reverse().join('.')
+
+            if (pathArray.length > longestResolvedPath.depth) {
+              longestResolvedPath.depth = pathArray.length
+              longestResolvedPath.formatted = formattedPath
+            }
+          }
+        }),
+        willSendResponse(responseContext) {
+          // default. only used if failed to parse
+          let transactionName = '*'
+
+          if (responseContext.operation) {
+            // Resolved operation: parsed and validated, will execute resolvers
+
+            const operationType = responseContext.operation.operation
+            const operationName = responseContext.operationName || '<anonymous>'
+            const longestPath = longestResolvedPath.formatted
+
+            transactionName = `${operationType} ${operationName} ${longestPath}`
+          } else if (responseContext.document) {
+            // Failed validation, fallback to document AST
+
+            transactionName = getNameFromDocument(responseContext.document)
+          }
+
+          setTransactionName(instrumentationApi.agent, transactionName)
+        }
+      }
+    }
+  }
+}
+
+function getNameFromDocument(document) {
+  // TODO: when would there be multiple document definitions?
+  const definition = document.definitions[0]
+  const operationType = definition.operation
+  const operationName = definition.name || '<anonymous>'
+
+  const longestSelectionPath = getLongestSelectionPath(definition)
+
+  const longestPath = longestSelectionPath.join('.')
+  return `${operationType} ${operationName} ${longestPath}`
+}
+
+function flattenToArray(fieldPath) {
+  const pathArray = []
+
+  let thisPath = fieldPath
+  while (thisPath) {
+    if (typeof thisPath.key !== 'number') {
+      pathArray.push(thisPath.key)
+    }
```

Comment:
Not sure if there's a faster way to strip the numbers here?

They seem to be representing the index of the resolver for nested resolvers but we don't really seem to have that on the document selections, and I don't think customers will find those numbers immediately obvious, so removed for consistency and clarity. Which also saves some UI space.